### PR TITLE
Feature/iroha main

### PR DESCRIPTION
**Pull request by lebdron**
## What is this pull request?
`irohad` transaction pipeline and `iroha-cli` transaction sending
  
## How to use the features provided in the pull request?
- Start `docker/run-ametsuchi-dev.sh`
- Create config file with following format
```JSON
{
  "block_store_path" : "/tmp/block_store/",
  "torii_port" : 50051,
  "pg_opt" : "host=192.168.9.4 port=5432 user=kek password=helloworld",
  "redis_host" : "192.168.9.3",
  "redis_port" : 6379
}
```
- Create genesis block
- Run `irohad --config .conf --genesis_block .block`
- Send transaction to `irohad`: `iroha-cli --grpc --json_transaction .json --address 192.168.9.2`

### CODE REVIEW

File: irohad/ametsuchi/block_serializer.hpp [hunk 1]
```diff
@@ -82,6 +82,9 @@ namespace iroha {
       nonstd::optional<model::Transaction> deserialize(
           const std::string json_tx);
 
+      void serialize(PrettyWriter<StringBuffer>& writer,
```
Comment (lebdron) [inline irohad/ametsuchi/block_serializer.hpp]: Move back to private functions

File: irohad/consensus/yac/impl/network_impl.cpp [hunk 2]
```diff
@@ -21,6 +21,22 @@
 
 using namespace logger;
 
+static const std::string code = {'0', '1', '2', '3', '4', '5', '6', '7',
+                                 '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
+
+std::string bytestringToHexstring(std::string str) {
```
Comment (lebdron) [inline irohad/consensus/yac/impl/network_impl.cpp]: Move to common utilitiess library

File: irohad/main/iroha_conf_loader.hpp [hunk 3]
```diff
@@ -47,22 +47,26 @@ inline rapidjson::Document parse_iroha_config(std::string const& iroha_conf_path
   doc.ParseStream(isw);
   assert_fatal(not doc.HasParseError(), "JSON parse error: " + iroha_conf_path);
 
-  assert_fatal(doc.HasMember(mbr::Ip), no_member_error(mbr::Ip));
+  /*assert_fatal(doc.HasMember(mbr::Ip), no_member_error(mbr::Ip));
```
Comment (lebdron) [inline irohad/main/iroha_conf_loader.hpp]: Remove, since it can be restored from commit history

File: irohad/model/model_hash_provider_impl.cpp [hunk 4]
```diff
@@ -23,6 +23,22 @@
 #include "model/queries/get_signatories.hpp"
 #include "model/queries/get_transactions.hpp"
 
+static const std::string code = {'0', '1', '2', '3', '4', '5', '6', '7',
+                                 '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
+
+std::string bytestringToHexstring1(std::string str) {
```
Comment (lebdron) [inline irohad/model/model_hash_provider_impl.cpp]: Remove

File: irohad/network/impl/peer_communication_service_impl.hpp [hunk 5]
```diff
@@ -14,38 +14,33 @@
...
 
       rxcpp::observable<Commit> on_commit() override;
 
      private:
-      network::OrderingGate &orderer_;
-      synchronizer::Synchronizer &synchronizer_;
+      std::shared_ptr<OrderingGate> ordering_gate_;
+      std::shared_ptr<synchronizer::Synchronizer> synchronizer_;
     };
-  }  // namespace network
```
Comment (lebdron) [inline irohad/network/impl/peer_communication_service_impl.hpp]: Add namespace comment

File: irohad/ordering/ordering_service_stub.cpp [hunk 6]
```diff
@@ -23,9 +23,8 @@ namespace iroha {
     using model::Transaction;
     using model::Proposal;
 
-    void OrderingGateStub::propagate_transaction(
```
Comment (lebdron) [inline irohad/ordering/ordering_service_stub.cpp]: Delete stub, since implementation is done

File: irohad/main/application.hpp [hunk 7]
```diff
@@ -17,34 +17,112 @@
...
+#include "simulator/impl/simulator.hpp"
+
+#include "main/impl/ordering_init.hpp"
+#include "main/impl/consensus_init.hpp"
 
 class Irohad {
  public:
-  std::shared_ptr<Context> context;
 
   Irohad(const std::string &block_store_dir, const std::string &redis_host,
```
Comment (grimadas) [inline irohad/main/application.hpp]: Add docs here for more clarity

**Pull request merged**